### PR TITLE
Add base58.ts and calle.ts

### PR DESCRIPTION
**Pull request by yusufidimaina9989**

### CODE REVIEW

**Comment by Harunadangabas** [comment]
Great work

File: src/contracts/calle.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import { assert } from "console"
+import { ByteString, SigHash, SmartContract, Utils, hash256, int2ByteString, method, prop } from "scrypt-ts"
+
+type Coeff = {
+  a : bigint
+  b : bigint
+  c : bigint
+}
+
+export class Calle extends SmartContract{
```
Comment (xhliu) [inline src/contracts/calle.ts:10]: Calle -> Callee

File: src/contracts/calle.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import { assert } from "console"
+import { ByteString, SigHash, SmartContract, Utils, hash256, int2ByteString, method, prop } from "scrypt-ts"
+
```
Comment (xhliu) [inline src/contracts/calle.ts:3]: For each contract, let us add a link to the article, if any, as in https://github.com/sCrypt-Inc/boilerplate/blob/b649ab516abb5eda8c2e442c1a752252abe02446/src/contracts/accumulatorMultiSig.ts#L14

https://xiaohuiliu.medium.com/inter-contract-call-on-bitcoin-f51869c08be

File: src/contracts/cointossxor.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import { assert } from "console";
+import { ByteString, PubKey, Sha256, Sig, SmartContract, hash256, len, method, prop, toByteString } from "scrypt-ts";
+
```
Comment (xhliu) [inline src/contracts/cointossxor.ts:3]: https://xiaohuiliu.medium.com/fair-bitcoin-toss-7310256e60c3

File: src/contracts/lottery.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
+// a fair secure multiparty lottery
+// each player i choose a random number n_i and the winner is the w-th player
+// where w = (n_0 + n_1 + ... + n_(N-1)) mod N
+
```
Comment (xhliu) [inline src/contracts/lottery.ts:4]: https://xiaohuiliu.medium.com/secure-multiparty-computations-on-bitcoin-953a64843b94

File: src/contracts/cointossxor.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        // we use 256 bits / 32 bytes as an example here
+
+        assert(BigInt(len(aliceNonce)) == 32n)
+        assert(hash256(aliceNonce) == this.aliceHash)
+        assert(BigInt(len(bobNonce) == 32n))
+        assert(hash256(bobNonce) == this.bobHash)
+
+        // last bit of XOR
+
+        const head : ByteString = (BigInt(aliceNonce) * BigInt(bobNonce)) && toByteString('0000000000000000000000000000000000000000000000000000000000000001')
```
Comment (xhliu) [inline src/contracts/cointossxor.ts:34]: `*` should be `xor` https://docs.scrypt.io/how-to-write-a-contract/built-ins#bitwise-operator

**Review by xhliu — CHANGES_REQUESTED** [review]
👍 A few minor changes.

File: src/contracts/Callee.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  @prop()
+  static readonly N : bigint = 2n
+
+  @method(SigHash.SINGLE)
+  public solve(co : Coeff, x : bigint){
+    assert(
+      co.a * x * x + co.b * x + co.c == 0n
+    )
+
+    const data : ByteString = int2ByteString(co.a, Calle.N) + int2ByteString(co.b, Calle.N) + int2ByteString(co.c, Calle.N)
```
Comment (zhfnjust) [inline src/contracts/Callee.ts:25]: `Calle` -> `Callee`

File: src/contracts/Callee.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
+
+
+import { assert } from "console"
```
Comment (zhfnjust) [inline src/contracts/Callee.ts:3]: import `assert` from `scrypt-ts`

File: src/contracts/cointossxor.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
+import { assert } from "console";
```
Comment (zhfnjust) [inline src/contracts/cointossxor.ts:1]: import assert from scrypt-ts

File: src/contracts/lottery.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
+// a fair secure multiparty lottery
+// each player i choose a random number n_i and the winner is the w-th player
+// where w = (n_0 + n_1 + ... + n_(N-1)) mod N
+
+// Read a Medium article about this contract
+//https://xiaohuiliu.medium.com/secure-multiparty-computations-on-bitcoin-953a64843b94
+
+import { assert } from "console";
```
Comment (zhfnjust) [inline src/contracts/lottery.ts:8]: import assert from scrypt-ts

File: src/contracts/Callee.ts (hunk 7 above)
Comment (yusufidimaina9989) [reply src/contracts/Callee.ts:3]: ok sir

**Pull request merged**